src/utils/price_comparison_utils.ts

```diff
@@ -0,0 +1,121 @@
+    ERC20BridgeSource.LiquidityProvider,
+]);
+
+const gasSchedule = {
+    [SwapVersion.V0]: GAS_SCHEDULE_V0,
+    [SwapVersion.V1]: GAS_SCHEDULE_V1,
+};
+
+const emptyPlaceholderSources = Object.values(ERC20BridgeSource)
+    .filter(liquiditySource => !excludedLiquiditySources.has(liquiditySource))
+    .reduce<SourceComparison[]>((memo, liquiditySource) => {
+        memo.push({
+            name: liquiditySource,
+            price: null,
+            gas: null,
+        });
+
+        return memo;
+    }, []);
+
+interface PartialRequestParams {
+    buyAmount?: BigNumber;
+    sellAmount?: BigNumber;
+}
+
+interface PartialQuote {
+    buyAmount: BigNumber;
+    sellAmount: BigNumber;
+    buyTokenAddress: string;
+    sellTokenAddress: string;
+    quoteReport?: { sourcesConsidered: QuoteReportSource[] };
+}
+
+export const priceComparisonUtils = {
+    getPriceComparisonFromQuote(
+        chainId: ChainId,
+        swapVersion: SwapVersion,
+        params: PartialRequestParams,
+        quote: PartialQuote,
+    ): SourceComparison[] | undefined {
+        if (!quote.quoteReport) {
+            logger.error('Missing quote report, cannot calculate price comparison');
+            return undefined;
+        }
+
+        // NOTE: don't fail quote request if comparison calculations error out
+        try {
+            const buyToken = getTokenMetadataIfExists(quote.buyTokenAddress, chainId);
+            const sellToken = getTokenMetadataIfExists(quote.sellTokenAddress, chainId);
+
+            const isSelling = !!params.sellAmount;
+
+            const { sourcesConsidered } = quote.quoteReport;
+
+            // Filter matching amount samples with a valid result
+            const fullTradeSources = sourcesConsidered.filter(s =>
+                isSelling
+                    ? s.takerAmount.isEqualTo(quote.sellAmount) && s.makerAmount.isGreaterThan(ZERO)
+                    : s.makerAmount.isEqualTo(quote.buyAmount) && s.takerAmount.isGreaterThan(ZERO),
+            );
+
+            // Strip out internal sources
+            const relevantSources = fullTradeSources.filter(s => !excludedLiquiditySources.has(s.liquiditySource));
+
+            // Sort by amount the user will receive and deduplicate to only take the best option for e.g. Kyber
+            const uniqueRelevantSources = _.uniqBy(
+                relevantSources
+                    .slice()
+                    .sort((a, b) =>
+                        isSelling ? b.makerAmount.comparedTo(a.makerAmount) : a.takerAmount.comparedTo(b.takerAmount),
+                    ),
+                'liquiditySource',
+            );
+
+            const sourcePrices: SourceComparison[] = uniqueRelevantSources.map(source => {
+                const { liquiditySource, makerAmount, takerAmount } = source;
+                const gas = new BigNumber(
+                    gasSchedule[swapVersion][source.liquiditySource]!((source as BridgeReportSource).fillData),
```

Comment:
Note v0 gas schedule is total gross including 0x overhead. (E.g 300k Uniswap)
v1 gas schedule is just the individual source cost estimate. (E.g 150k Uniswap where we add the base cost in later, 150 Uni  + 150 Base)

Reply:
Good point! It probably makes more sense to always use v1 gas schedule as it is the most accurate representation of the gas cost for each source independently of 0x overhead 👍